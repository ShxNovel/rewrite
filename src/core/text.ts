export interface TextAddonType {}
export interface RewriteTextType {
    pause: never;
    instant: never;
    until: never;
}

export type RewriteText =
    | string
    | {
          type: keyof RewriteTextType | keyof TextAddonType;
          args: Record<PropertyKey, unknown>;
      };

export interface Text {
    type: 'text';
    name: string | null;
    content: RewriteText[];
}

/** @param ms The duration of the pause in milliseconds */
export function pause(ms: number): RewriteText {
    return { type: 'pause', args: { ms } };
}

/** @param str The string content to be displayed */
export function instant(str: string): RewriteText {
    return { type: 'instant', args: { str } };
}

export function until(str: string | null): RewriteText {
    return { type: 'until', args: { str } };
}
