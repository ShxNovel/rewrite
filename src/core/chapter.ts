import { cache } from './cache';
import { makeEffect, Effect, EffectMethods } from './effect';
import { RewriteText, Text } from './text';

export type ChapterUnit = Text | Effect;

export type LinkText = (some: TemplateStringsArray, ...values: RewriteText[]) => LinkText;

export type ChapterInfo = {
    memory: ChapterUnit[];
};

export function chapter() {
    const memory: ChapterUnit[] = [];

    const context: ChapterInfo = { memory };

    cache.push(context);

    class Chapter {
        constructor() {}

        character(name: string | null): LinkText {
            function init(some: TemplateStringsArray, ...values: RewriteText[]) {
                const content: RewriteText[] = [];
                const talk: Text = { type: 'text', name, content };

                memory.push(talk);

                function link(some: TemplateStringsArray, ...values: RewriteText[]): LinkText {
                    const len = some.length;
                    content.push(some[0]);
                    for (let i = 1; i < len; i++) {
                        content.push(values[i - 1]);
                        content.push(some[i]);
                    }
                    return link;
                }
                return link(some, ...values);
            }
            return init;
        }

        effect(): EffectMethods {
            return makeEffect(context);
        }

        EndOfChapter(): ChapterInfo {
            return context;
        }
    }

    return new Chapter();
}
