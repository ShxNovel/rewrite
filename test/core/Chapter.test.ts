import { assert, expectTypeOf, assertType, expect, test, describe } from 'vitest';
import { chapter } from '../../src/core/chapter.ts';
import { pause, RewriteText } from '../../src/core/text.ts';

describe('text', () => {
    test('RewriteText', () => {
        function say() {
            return { a: 1 };
        }

        // @ts-expect-error
        assertType<RewriteText>(say());
    });
});

describe('Chapter', () => {
    test('type', () => {
        const file = chapter();

        const { character, EndOfChapter } = file;
        const x = character('aside'); // 旁白
        const you = character('hero'); // 你

        function say() {
            return { a: 1 };
        }

        // @ts-expect-error
        x`some${say()}``text```;
        // @ts-expect-error
        you`some${say()}``text```;

        const res = EndOfChapter();
        expect(res.memory).toMatchSnapshot();
    });

    test('content', () => {
        const file = chapter();
        const { character, EndOfChapter } = file;

        const x = character('aside');
        const you = character('hero');

        x`some``text```;
        you`some``text```;
        x`some``text```;

        const res = EndOfChapter();
        expect(res.memory).toMatchSnapshot();
    });

    test('pause', () => {
        const file = chapter();
        const { character, EndOfChapter } = file;

        const x = character('aside');
        const you = character('hero');

        x`some${pause(10)}``${pause(10)}${pause(10)}text${pause(10)}``${pause(10)}`;
        you`some``text```;

        const res = EndOfChapter();
        expect(res.memory).toMatchSnapshot();
    });

    test('aim', () => {
        const { effect, EndOfChapter } = chapter();

        effect().aim('some').move();

        // @ts-expect-error
        effect().move();

        const res = EndOfChapter();
        expect(res).toMatchSnapshot();
    });
});
