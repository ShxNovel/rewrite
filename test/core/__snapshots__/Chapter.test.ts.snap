// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chapter > aim 1`] = `
{
  "memory": [
    {
      "content": [
        {
          "args": {
            "name": "some",
            "scene": "main",
          },
          "type": "aim",
        },
        {
          "args": {},
          "type": "move",
        },
      ],
      "type": "motion",
    },
    {
      "content": [
        {
          "args": {},
          "type": "move",
        },
      ],
      "type": "motion",
    },
  ],
}
`;

exports[`Chapter > content 1`] = `
[
  {
    "content": [
      "some",
      "text",
      "",
    ],
    "name": "aside",
    "type": "text",
  },
  {
    "content": [
      "some",
      "text",
      "",
    ],
    "name": "hero",
    "type": "text",
  },
  {
    "content": [
      "some",
      "text",
      "",
    ],
    "name": "aside",
    "type": "text",
  },
]
`;

exports[`Chapter > pause 1`] = `
[
  {
    "content": [
      "some",
      {
        "args": {
          "ms": 10,
        },
        "type": "pause",
      },
      "",
      "",
      {
        "args": {
          "ms": 10,
        },
        "type": "pause",
      },
      "",
      {
        "args": {
          "ms": 10,
        },
        "type": "pause",
      },
      "text",
      {
        "args": {
          "ms": 10,
        },
        "type": "pause",
      },
      "",
      "",
      {
        "args": {
          "ms": 10,
        },
        "type": "pause",
      },
      "",
    ],
    "name": "aside",
    "type": "text",
  },
  {
    "content": [
      "some",
      "text",
      "",
    ],
    "name": "hero",
    "type": "text",
  },
]
`;

exports[`Chapter > type 1`] = `
[
  {
    "content": [
      "some",
      {
        "a": 1,
      },
      "",
      "text",
      "",
    ],
    "name": "aside",
    "type": "text",
  },
  {
    "content": [
      "some",
      {
        "a": 1,
      },
      "",
      "text",
      "",
    ],
    "name": "hero",
    "type": "text",
  },
]
`;
